### PR TITLE
[Feature]Drop lake table 

### PR DESCRIPTION
**Pull request by abc982627271**
## What type of PR is this：
- [ ] bug
- [x] feature
- [ ] enhancement
- [ ] refactor
- [ ] others

## Which issues of this PR fixes ：

Fixes #

## Problem Summary(Required) ：

drop lake table:
1.use a daemon thread to drop lake tablet and delete shard
2.persist shard infos in image and journal

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 58 / 171 (33.92%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/StarOSAgent.java|0|19|00.00%|
|:large_blue_circle:|com/starrocks/server/ShardManager.java|0|14|00.00%|
|:large_blue_circle:|com/starrocks/journal/JournalEntity.java|0|6|00.00%|
|:large_blue_circle:|com/starrocks/persist/EditLog.java|2|10|20.00%|
|:large_blue_circle:|com/starrocks/catalog/lake/ShardDeleter.java|15|70|21.43%|
|:large_blue_circle:|com/starrocks/catalog/CatalogRecycleBin.java|2|6|33.33%|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|3|7|42.86%|
|:large_blue_circle:|com/starrocks/catalog/Database.java|3|4|75.00%|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|7|9|77.78%|
|:large_blue_circle:|com/starrocks/persist/ShardInfo.java|12|12|100.00%|
|:large_blue_circle:|com/starrocks/catalog/lake/ShardManager.java|14|14|100.00%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 64 / 147 (43.54%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/journal/JournalEntity.java|0|6|00.00%|
|:large_blue_circle:|com/starrocks/persist/EditLog.java|2|10|20.00%|
|:large_blue_circle:|com/starrocks/catalog/lake/ShardDeleter.java|15|70|21.43%|
|:large_blue_circle:|com/starrocks/catalog/CatalogRecycleBin.java|2|6|33.33%|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|3|7|42.86%|
|:large_blue_circle:|com/starrocks/catalog/StarOSAgent.java|6|9|66.67%|
|:large_blue_circle:|com/starrocks/catalog/Database.java|3|4|75.00%|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|7|9|77.78%|
|:large_blue_circle:|com/starrocks/persist/ShardInfo.java|12|12|100.00%|
|:large_blue_circle:|com/starrocks/catalog/lake/ShardManager.java|14|14|100.00%|

File: gensrc/proto/lake_service.proto [hunk 1]
```diff
@@ -29,6 +29,15 @@ message AbortTxnResponse {
     repeated int64 failed_tablets = 1;
 }
 
+message DropTabletRequest {
+    repeated int64 tablet_ids = 1;
+}
+
+message DropTabletResponse {
+    repeated int64 succ_tablets = 2;
```
Comment (wyb) [inline gensrc/proto/lake_service.proto:37]: Better to return failed tablets because most cases are normal.

Comment (wyb) [inline gensrc/proto/lake_service.proto:37]:
```suggestion
    repeated int64 succ_tablets = 1;
```

File: fe/fe-core/src/main/java/com/starrocks/server/LocalMetastore.java [hunk 2]
```diff
@@ -4160,7 +4165,19 @@ public HashMap<Long, AgentBatchTask> onEraseOlapTable(OlapTable olapTable, boole
...
+                            // for debug
+                            LOG.info("onEraseOlapTable, isLakeTable");
+                            tabletIds.add(tabletId);
+                        }
+                        
                     } // end for tablets
+
+                    GlobalStateMgr.getCurrentState().getShardManager()
+                            .getShardDeleter().addUnusedShardId(tabletIds);
+                    GlobalStateMgr.getCurrentState().getEditLog().logAddDeleteShard(tabletIds);
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/server/LocalMetastore.java:4180]: Move to outer loop and execute `add` and `log` on table level.

Comment (wyb) [reply fe/fe-core/src/main/java/com/starrocks/server/LocalMetastore.java:4180]: Should check if table is lake table.

File: fe/fe-core/src/main/java/com/starrocks/catalog/lake/ShardDeleter.java [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+                try {
+                    DropTabletResponse response = responseFuture.get();
+                    if (response == null || response.succTablets == null || response.succTablets.isEmpty()) {
+                        continue;
+                    } else {
+                        // for debug
+                        LOG.info("succTablets is {}", response.succTablets);
+                        shards = new HashSet<>(response.succTablets);
+                    }
+                } catch (ExecutionException | InterruptedException e) {
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/catalog/lake/ShardDeleter.java:106]:
```suggestion
                } catch (Exception e) {
```

File: fe/fe-core/src/main/java/com/starrocks/catalog/lake/ShardDeleter.java [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+                    LOG.error(e);
+                    finished = false;
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/catalog/lake/ShardDeleter.java:108]: continue;

File: fe/fe-core/src/main/java/com/starrocks/catalog/lake/ShardDeleter.java [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+                }
+            }
+
+            // 2. delete shard
+            try {
+                GlobalStateMgr.getCurrentState().getStarOSAgent().deleteShards(shards);
+                // for debug
+                LOG.info("delete shards {} succ.", shards);
+            } catch (DdlException e) {
+                LOG.warn("failed to delete shard from starMgr");
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/catalog/lake/ShardDeleter.java:118]: finished = false;

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by abc982627271** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 112 / 143 (78.32%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/journal/JournalEntity.java|0|6|00.00%|
|:large_blue_circle:|com/starrocks/catalog/CatalogRecycleBin.java|2|5|40.00%|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|3|7|42.86%|
|:large_blue_circle:|com/starrocks/catalog/StarOSAgent.java|6|9|66.67%|
|:large_blue_circle:|com/starrocks/catalog/Database.java|3|4|75.00%|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|7|9|77.78%|
|:large_blue_circle:|com/starrocks/persist/EditLog.java|8|10|80.00%|
|:large_blue_circle:|com/starrocks/catalog/lake/ShardDeleter.java|57|67|85.07%|
|:large_blue_circle:|com/starrocks/persist/ShardInfo.java|12|12|100.00%|
|:large_blue_circle:|com/starrocks/catalog/lake/ShardManager.java|14|14|100.00%|

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by abc982627271** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 114 / 144 (79.17%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/CatalogRecycleBin.java|2|5|40.00%|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|3|7|42.86%|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|5|10|50.00%|
|:large_blue_circle:|com/starrocks/persist/EditLog.java|6|10|60.00%|
|:large_blue_circle:|com/starrocks/catalog/StarOSAgent.java|6|9|66.67%|
|:large_blue_circle:|com/starrocks/catalog/Database.java|3|4|75.00%|
|:large_blue_circle:|com/starrocks/catalog/lake/ShardDeleter.java|57|67|85.07%|
|:large_blue_circle:|com/starrocks/persist/ShardInfo.java|12|12|100.00%|
|:large_blue_circle:|com/starrocks/catalog/lake/ShardManager.java|14|14|100.00%|
|:large_blue_circle:|com/starrocks/journal/JournalEntity.java|6|6|100.00%|

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

File: fe/fe-core/src/main/java/com/starrocks/catalog/StarOSAgent.java [hunk 6]
```diff
@@ -244,6 +244,18 @@ public long getWorkerIdByBackendId(long backendId) {
         }
     }
 
+    public void deleteShards(Set<Long> shardIds) throws DdlException {
+        prepare();
+        try {
+            client.deleteShard(serviceId, shardIds);
+        } catch (StarClientException e) {
+            if (e.getCode() != StatusCode.NOT_EXIST) {
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/catalog/StarOSAgent.java:252]: Remove this check.

File: fe/fe-core/src/main/java/com/starrocks/persist/OperationType.java [hunk 7]
```diff
@@ -243,4 +243,8 @@ public class OperationType {
     // manage compute node 10201 ~ 10220
     public static final short OP_ADD_COMPUTE_NODE = 10201;
     public static final short OP_DROP_COMPUTE_NODE = 10202;
+
+    // shard operate 10221 ~ 10222
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/persist/OperationType.java:247]:
```suggestion
    // shard operate 10221 ~ 10240
```

File: fe/fe-core/src/main/java/com/starrocks/catalog/CatalogRecycleBin.java [hunk 8]
```diff
@@ -328,8 +329,9 @@ public synchronized void replayEraseTable(long tableId) {
             RecycleTableInfo tableInfo = idToTableInfo.remove(dbId, tableId);
             if (tableInfo != null) {
                 Table table = tableInfo.getTable();
+                TableType tableType = table.getType();
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/catalog/CatalogRecycleBin.java:332]: Not used.

File: fe/fe-core/src/main/java/com/starrocks/catalog/lake/ShardDeleter.java [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+        try (LockCloseable lock = new LockCloseable(rwLock.writeLock())) {
+            shardIds.addAll(tableIds);
+        }
+    }
+
+    private void deleteUnusedShard() {
+        // delete shard and drop lakeTablet
+        if (shardIds.isEmpty()) {
+            // for debug
+            LOG.info("shardIds is empty");
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/catalog/lake/ShardDeleter.java:55]:
```suggestion
            LOG.debug("shardIds is empty");
```

File: fe/fe-core/src/main/java/com/starrocks/catalog/lake/ShardDeleter.java [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+            } catch (DdlException e) {
+                LOG.warn("failed to delete shard from starMgr");
+                continue;
+            }
+
+            // 3. succ both, remove from the map
+            GlobalStateMgr.getCurrentState().getEditLog().logDeleteUnusedShard(shards);
+            try (LockCloseable lock = new LockCloseable(rwLock.writeLock())) {
+                shardIds.removeAll(shards);
+            }
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/catalog/lake/ShardDeleter.java:117]: Better to log once after all BEs are processed.

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 105 / 133 (78.95%)




#### file detail

|   |path|covered line|new line|coverage|not covered line detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/CatalogRecycleBin.java|2|5|40.00%|[748, 752, 753]|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|3|7|42.86%|[4193, 4201, 4202, 4203]|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|3|7|42.86%|[1148, 1149, 1400, 1401]|
|:large_blue_circle:|com/starrocks/persist/EditLog.java|6|10|60.00%|[1503, 1504, 1507, 1508]|
|:large_blue_circle:|com/starrocks/catalog/StarOSAgent.java|5|8|62.50%|[250, 256, 257]|
|:large_blue_circle:|com/starrocks/catalog/lake/ShardDeleter.java|50|60|83.33%|[65, 66, 93, 94, 97, 98, 99, 106, 107, 108]|
|:large_blue_circle:|com/starrocks/catalog/Database.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/persist/ShardInfo.java|12|12|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/lake/ShardManager.java|14|14|100.00%|[]|
|:large_blue_circle:|com/starrocks/journal/JournalEntity.java|6|6|100.00%|[]|

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 105 / 133 (78.95%)




#### file detail

|   |path|covered line|new line|coverage|not covered line detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/CatalogRecycleBin.java|2|5|40.00%|[748, 752, 753]|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|3|7|42.86%|[4193, 4201, 4202, 4203]|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|3|7|42.86%|[1148, 1149, 1400, 1401]|
|:large_blue_circle:|com/starrocks/persist/EditLog.java|6|10|60.00%|[1503, 1504, 1507, 1508]|
|:large_blue_circle:|com/starrocks/catalog/StarOSAgent.java|5|8|62.50%|[250, 256, 257]|
|:large_blue_circle:|com/starrocks/catalog/lake/ShardDeleter.java|50|60|83.33%|[65, 66, 93, 94, 97, 98, 99, 106, 107, 108]|
|:large_blue_circle:|com/starrocks/catalog/Database.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/persist/ShardInfo.java|12|12|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/lake/ShardManager.java|14|14|100.00%|[]|
|:large_blue_circle:|com/starrocks/journal/JournalEntity.java|6|6|100.00%|[]|

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 106 / 137 (77.37%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/CatalogRecycleBin.java|1|6|16.67%|[380, 747, 749, 751, 752]|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|3|7|42.86%|[4193, 4201, 4202, 4203]|
|:large_blue_circle:|com/starrocks/catalog/StarOSAgent.java|4|7|57.14%|[250, 254, 255]|
|:large_blue_circle:|com/starrocks/persist/EditLog.java|6|10|60.00%|[1503, 1504, 1507, 1508]|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|7|11|63.64%|[1148, 1149, 1406, 1407]|
|:large_blue_circle:|com/starrocks/catalog/Database.java|4|5|80.00%|[407]|
|:large_blue_circle:|com/starrocks/catalog/lake/ShardDeleter.java|51|61|83.61%|[65, 66, 95, 96, 99, 100, 101, 108, 109, 110]|
|:large_blue_circle:|com/starrocks/persist/ShardInfo.java|12|12|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/lake/ShardManager.java|12|12|100.00%|[]|
|:large_blue_circle:|com/starrocks/journal/JournalEntity.java|6|6|100.00%|[]|

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Review by wyb — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 106 / 137 (77.37%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/CatalogRecycleBin.java|1|6|16.67%|[380, 747, 749, 751, 752]|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|3|7|42.86%|[4193, 4201, 4202, 4203]|
|:large_blue_circle:|com/starrocks/catalog/StarOSAgent.java|4|7|57.14%|[250, 254, 255]|
|:large_blue_circle:|com/starrocks/persist/EditLog.java|6|10|60.00%|[1503, 1504, 1507, 1508]|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|7|11|63.64%|[1148, 1149, 1406, 1407]|
|:large_blue_circle:|com/starrocks/catalog/Database.java|4|5|80.00%|[407]|
|:large_blue_circle:|com/starrocks/catalog/lake/ShardDeleter.java|51|61|83.61%|[65, 66, 95, 96, 99, 100, 101, 108, 109, 110]|
|:large_blue_circle:|com/starrocks/persist/ShardInfo.java|12|12|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/lake/ShardManager.java|12|12|100.00%|[]|
|:large_blue_circle:|com/starrocks/journal/JournalEntity.java|6|6|100.00%|[]|

**Review by gengjun-git — APPROVED** [review]

**Comment by wyb** [comment]
run starrocks_admit_test

**Review by wyb — APPROVED** [review]

**Pull request merged**